### PR TITLE
create automated alerts from jobs

### PR DESCRIPTION
**Pull request by franzns**
- All jobs are now scheduled via manual-jobs
- All jobs publish a `-done` metric when they ran without an error
- When scheduling jobs, an alarm is also created on cloudwatch
- The alarm is set to go off if the cron did not run once in the doubled interval
   - e.g. a cron runs every minute, alarm is triggered if the crons runs less or equal than once every two minutes
   - very short intervals (<30s) will be checked to run once per minute

- Small fix in the scheduling to track sentry transactions correctly on error.

### CODE REVIEW

**Comment by franzns** [comment]
addresses #58

File: worker/create-alerts.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    DescribeAlarmsCommand,
+    PutMetricAlarmCommand,
+} from '@aws-sdk/client-cloudwatch';
+import { env } from '../app/env';
+import cronRunMetric from '../modules/metrics/cron.metric';
+import { WorkerJob } from './manual-jobs';
+
+export async function createAlertsIfNotExist(jobs: WorkerJob[]): Promise<void> {
+    const cloudWatchClient = new CloudWatchClient({
+        region: process.env.AWS_REGION,
```
Comment (franzns) [inline worker/create-alerts.ts:13]: This is an assumption that process.env.AWS_REGION will always be present by default. Correct?

File: worker/worker.ts [hunk 2]
```diff
@@ -35,7 +35,8 @@ export function startWorker() {
         }),
     );
 
-    scheduleManualJobs();
+    scheduleJobs();
+    createAlerts();
```
Comment (0xSkly) [inline worker/worker.ts:39]: we could also think about running this on CI ? but also fine here

File: worker/manual-jobs.ts [hunk 3]
```diff
@@ -1,19 +1,30 @@
...
-async function scheduleJobWithInterval(job: WokerJobType, intervalMs: number): Promise<void> {
-    await workerQueue.sendWithInterval(JSON.stringify({ type: job }), intervalMs);
+async function scheduleJobWithInterval(job: string, intervalMs: number): Promise<void> {
+    await workerQueue.sendWithInterval(JSON.stringify({ name: job }), intervalMs);
 }
 
-// all jobs requiring manual scheduling will be handled here (e.g. sub minute crons)
+export async function scheduleJobs(): Promise<void> {
+    if (isFantomNetwork()) {
+        fantomJobs.forEach(async (job) => await scheduleJobWithInterval(job.name, job.interval));
```
Comment (0xSkly) [inline worker/manual-jobs.ts:18]: this works, but might not exactly do what you expect, read this https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

File: worker/create-alerts.ts (hunk 1 above)
Comment (0xSkly) [reply worker/create-alerts.ts:14]: it has to be manually put into the environment variables, its not injected by AWS or something if you mean this

File: worker/intervals.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        case Duration.Hours:
+            const msInHour = 3_600_000;
+            return msInHour * times;
+        case Duration.Days:
+            const msInDay = 86_400_000;
+            return msInDay * times;
+        default:
+            throw new Error(`Unknown duration: ${duration}`);
+    }
+}
```
[the remark this answers is not in the document]
Comment (franzns) [reply worker/intervals.ts:28]: Wow, true. Why are there enums in typescript? Always forget I can just create simple "string" types...

File: worker/create-alerts.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        //make sure metric is available for alarm
+        cronRunMetric.publish(`${cronJob.name}-done`);
+
+        const putAlarmCommand = new PutMetricAlarmCommand({
+            AlarmName: alarmName,
+            AlarmDescription: `The cron job ${cronJob.name} should run every ${cronJob.interval / 1000} seconds. 
+            Trigger alarm if the cron ran less than once in ${periodInSeconds} seconds.`,
+            ActionsEnabled: true,
+            AlarmActions: ['arn:aws:sns:eu-central-1:837533371577:Default_CloudWatch_Alarms_Topic'],
```
[the remark this answers is not in the document]
Comment (franzns) [reply worker/create-alerts.ts:59]: Ha, good catch! Didn't think about that one

File: worker/worker.ts (hunk 2 above)
Comment (franzns) [reply worker/worker.ts:39]: Yeah true. There are other things we should consider to put into a CI e.g. stellate config.

File: modules/metrics/metrics.client.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    PutMetricDataCommand,
+} from '@aws-sdk/client-cloudwatch';
+import { env } from '../../app/env';
+
+export interface NotificationsCloudwatchMetric {
+    merticData: MetricDatum[];
+    nameSpace: string;
+}
+
+export abstract class MetricPublisher {
```
[the remark this answers is not in the document]
Comment (franzns) [reply modules/metrics/metrics.client.ts]: legacy from before

File: modules/metrics/metrics.client.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                        Timestamp: new Date(),
+                        Value: count ? count : 1,
+                    },
+                ],
+                Namespace: this.namespace,
+            });
+
+            await this.client.send(command);
+        } catch (err) {
+            console.log('Failed to publish metric ', err);
```
[the remark this answers is not in the document]
Comment (franzns) [reply modules/metrics/metrics.client.ts]: nope, logging it now.

File: modules/metrics/metrics.client.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+export abstract class MetricPublisher {
+    protected environment: string;
+    protected configuration?: CloudWatchClientConfig;
+    private client: CloudWatchClient;
+    public namespace: string;
+
+    constructor(configuration?: CloudWatchClientConfig) {
+        this.namespace = 'default';
+        this.environment = env.DEPLOYMENT_ENV;
```
[the remark this answers is not in the document]
Comment (franzns) [reply modules/metrics/metrics.client.ts]: refactored the whole thing

File: package.json [hunk 9]
```diff
@@ -40,6 +41,7 @@
         "graphql-request": "^3.6.1",
         "graphql-type-json": "^0.3.2",
         "helmet": "^5.0.2",
+        "js-yaml": "^4.1.0",
```
[the remark this answers is not in the document]
Comment (franzns) [reply package.json]: removed

**Review by 0xSkly — APPROVED** [review]
perfect 👍

File: worker/intervals.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xSkly) [reply worker/intervals.ts]: yea guess there are some use cases where they make more sense? not sure,  also preference i guess

**Pull request merged**